analyzer/src/main/kotlin/managers/Pip.kt

```diff
@@ -679,7 +679,7 @@ class Pip(
         }
 
         val declaredLicenses = sortedSetOf<String>()
-        getLicenseFromLicenseField(map["License"]?.single())?.let { declaredLicenses += it }
+        map["License"]?.mapNotNullTo(declaredLicenses) { getLicenseFromLicenseField(it) }
```

Comment:
commit: Do you have a reference to the case (package)  where this problem appeared?

Reply:
No. According to @mmurto's report in the chat it should be one of  `dill Django psycopg2 pymongo python-dateutil pytz PyYAML sqlparse` (unclear in which version), but I cannot spend the time to find out exactly now.

Reply:
I was just wondering whether the issue could be a bug in lines 660 - 674. Maybe we just wait for @mmurto 's test results.

Reply:
See https://ort-talk.slack.com/archives/C9NNJ54B1/p1623930025217500?thread_ts=1623924964.216000&cid=C9NNJ54B1 for @mmurto's feedback. So given it's a clear misuse of the "License" field, should I drop this change again?

Reply:
Or maybe we should not fail hard but only warn if the "License" contains multiple fields?

Reply:
The correct use `license` argument seems to be [somewhat unclear](https://github.com/pypa/packaging.python.org/issues/95).  There's also some discussion at [here](https://discuss.python.org/t/pep-639-improving-license-clarity-with-better-package-metadata/2154). Haven't read through all of this and all the linked issues, but it seems that currently the `license` argument is something alike "free form text about licencing". [One PR](https://github.com/pypa/packaging.python.org/pull/753) removed some clarifications that would've pointed to it being a more structured field.

Reply:
The analyzer run succeeded with the following result:

```
    - id: "PIP::<redacted>:5743024ffd8a42b806a13060104d861146d4e5ba"
      definition_file_path: "<redacted>/requirements.txt"
      authors:
      - "<redacted>"
      declared_licenses:
      - "\nThe MIT License \nCopyright (c) 2021- <redacted>\
        \ <redacted>\nCopyright (c) 2017-2020 <redacted>\
        \ (<redacted>)\n \nPermission is hereby granted, free of charge, to any person obtaining\
        \ a copy \nof this software and associated documentation files (the \"Software\"\
        ), to deal \nin the Software without restriction, including without limitation\
        \ the rights \nto use, copy, modify, merge, publish, distribute, sublicense,\
        \ and/or sell \ncopies of the Software, and to permit persons to whom the\
        \ Software is \nfurnished to do so, subject to the following conditions: \n\
        \ \nThe above copyright notice and this permission notice shall be included\
        \ in \nall copies or substantial portions of the Software. \n \nTHE SOFTWARE\
        \ IS PROVIDED \"AS IS\", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR \nIMPLIED,\
        \ INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, \nFITNESS\
        \ FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE \nAUTHORS\
        \ OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER \nLIABILITY,\
        \ WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, \nOUT\
        \ OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN \n\
        THE SOFTWARE."
      - "MIT License"
      declared_licenses_processed:
        spdx_expression: "MIT"
        mapped:
          MIT License: "MIT"
        unmapped:
        - "\nThe MIT License \nCopyright (c) 2021- <redacted>\
          \ <redacted> \nCopyright (c) 2017-2020 <redacted>\
          \ <redacted>)\n \nPermission is hereby granted, free of charge, to any\
          \ person obtaining a copy \nof this software and associated documentation\
          \ files (the \"Software\"), to deal \nin the Software without restriction,\
          \ including without limitation the rights \nto use, copy, modify, merge,\
          \ publish, distribute, sublicense, and/or sell \ncopies of the Software,\
          \ and to permit persons to whom the Software is \nfurnished to do so, subject\
          \ to the following conditions: \n \nThe above copyright notice and this\
          \ permission notice shall be included in \nall copies or substantial portions\
          \ of the Software. \n \nTHE SOFTWARE IS PROVIDED \"AS IS\", WITHOUT WARRANTY\

```

Reply:
@sschuberth how about changing the logic to ignore any "license" property values which are not single line?

Reply:
If by "not single line" you mean "the string does not contain a newline character" then this should actually work, or? I'll give it a try.

Reply:
I've added a commit. Please have a look @fviernau. And @mmurto you're welcome to test once more 😉

Reply:
Analyzer succeeds, but declared contain it all:

```
        declared_licenses:
        - "AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER"
        - <redacted>
        - "FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL\
          \ THE"
        - "IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,"
        - "LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING\
          \ FROM,"
        - "MIT License"
        - "OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS\
          \ IN"
        - "Permission is hereby granted, free of charge, to any person obtaining a\
          \ copy"
        - "THE SOFTWARE IS PROVIDED \"AS IS\", WITHOUT WARRANTY OF ANY KIND, EXPRESS\
          \ OR"
        - "THE SOFTWARE."
        - "The MIT License"
        - "The above copyright notice and this permission notice shall be included\
          \ in"
        - "all copies or substantial portions of the Software."
        - "copies of the Software, and to permit persons to whom the Software is"
        - "furnished to do so, subject to the following conditions:"
        - "in the Software without restriction, including without limitation the rights"
        - "of this software and associated documentation files (the \"Software\"),\
          \ to deal"
        - "to use, copy, modify, merge, publish, distribute, sublicense, and/or sell"
        declared_licenses_processed:
          spdx_expression: "MIT"
          mapped:
            MIT License: "MIT"
            The MIT License: "MIT"
          unmapped:
          - "AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER"
          - <redacted>
          - "FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL\
            \ THE"
          - "IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,"
          - "LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING\
            \ FROM,"
          - "OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS\
            \ IN"
          - "Permission is hereby granted, free of charge, to any person obtaining\
            \ a copy"
          - "THE SOFTWARE IS PROVIDED \"AS IS\", WITHOUT WARRANTY OF ANY KIND, EXPRESS\
            \ OR"
          - "THE SOFTWARE."
          - "The above copyright notice and this permission notice shall be included\
            \ in"
          - "all copies or substantial portions of the Software."
          - "copies of the Software, and to permit persons to whom the Software is"
          - "furnished to do so, subject to the following conditions:"
          - "in the Software without restriction, including without limitation the\
            \ rights"
          - "of this software and associated documentation files (the \"Software\"\
            ), to deal"
          - "to use, copy, modify, merge, publish, distribute, sublicense, and/or\
            \ sell"

```

Reply:
> The analyzer run succeeded with the following result:
> 
> ```
>     - id: "PIP::<redacted>:5743024ffd8a42b806a13060104d861146d4e5ba"
>       definition_file_path: "<redacted>/requirements.txt"
>       authors:
>       - "<redacted>"
>       declared_licenses:
>       - "\nThe MIT License \nCopyright (c) 2021- <redacted>\
> ```

@mmurto, is the `"<redacted>/requirements.txt"` part in here really correct? Shouldn't it be `"<redacted>/setup.py"`?

Because [above](https://github.com/oss-review-toolkit/ort/pull/4184#discussion_r653487908) you wrote

> the packages has the whole license text of MIT in setup.py

and `requirements.txt` does not actually support declaring a license AFAIK.

Reply:
> > The analyzer run succeeded with the following result:
> > ```
> >     - id: "PIP::<redacted>:5743024ffd8a42b806a13060104d861146d4e5ba"
> >       definition_file_path: "<redacted>/requirements.txt"
> >       authors:
> >       - "<redacted>"
> >       declared_licenses:
> >       - "\nThe MIT License \nCopyright (c) 2021- <redacted>\
> > ```
> 
> @mmurto, is the `"<redacted>/requirements.txt"` part in here really correct? Shouldn't it be `"<redacted>/setup.py"`?
> 
> Because [above](https://github.com/oss-review-toolkit/ort/pull/4184#discussion_r653487908) you wrote
> 
> > the packages has the whole license text of MIT in setup.py
> 
> and `requirements.txt` does not actually support declaring a license AFAIK.

The part with `requirements.txt` is correct, the project's `requirements.txt` contains only `.`, which leads to pip getting the dependencies from `setup.py`, [some info here](https://stackoverflow.com/questions/14399534/reference-requirements-txt-for-the-install-requires-kwarg-in-setuptools-setup-py).

Reply:
> The part with `requirements.txt` is correct, the project's `requirements.txt` contains only `.`, which leads to pip getting the dependencies from `setup.py`, [some info here](https://stackoverflow.com/questions/14399534/reference-requirements-txt-for-the-install-requires-kwarg-in-setuptools-setup-py).

Wow, subtle!